inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp

```diff
@@ -2342,3 +2298,81 @@ void MKLDNNGraphOptimizer::MergePermuteAndReorder(MKLDNNGraph &graph) {
+    auto status = mkldnn_reorder_primitive_desc_create_v2(&result, srcPrimitiveDesc.get(), dstPrimitiveDesc.get(), attr.get());
+    if (result) {
+        mkldnn_primitive_desc_destroy(result);
+    }
+
+    return mkldnn_success == status;
+}
+
+void MKLDNNGraphOptimizer::AddConvertToReorder(MKLDNNGraph &graph) {
+    const auto& vecNodes = graph.GetNodes();
+    std::set<MKLDNNNodePtr> insertedReorders;
+
+    for (int i = 0; i < vecNodes.size(); i++) {
+        auto node = vecNodes[i];
+        if (Reorder == node->getType()) {
+            auto inpEdge = node->getParentEdgeAt(0);
+            auto outEdge = node->getChildEdgeAt(0);
+            // Some guard checks
+            if (inpEdge->getDesc().getPrecision() == outEdge->getDesc().getPrecision()) {
+                continue;
+            }
+
+            if (insertedReorders.count(node)) {
+                continue;
+            }
+
+            // Check if there is a reorder that supports the type conversion
+            if (isReorderAvailable(inpEdge, outEdge, graph.getEngine())) {
+                continue;
+            }
+
+            std::string convertName = inpEdge->getParent()->getName() + "_" +
+                    inpEdge->getDesc().getPrecision().name() + "_" + outEdge->getDesc().getPrecision().name();
+            //Insert Convert before the reorder
+            // make CNNLayer
+            CNNLayerPtr convert(new  CNNLayer(LayerParams{convertName, "Convert", outEdge->getDesc().getPrecision()}));
+
+            auto oIndex = outEdge->getOutputNum();
+            auto iIndex = inpEdge->getInputNum();
+            if (iIndex < 0 || oIndex < 0)
+                THROW_IE_EXCEPTION << "Cannot insert node '" << node->getName() << "' between nodes: "
+                                   << inpEdge->getParent()->getName() << " and "
+                                   << outEdge->getChild()->getName() << ".";
+
+            auto convertNode = std::make_shared<MKLDNNConvertNode>(convert, graph.getEngine(), graph.weightsCache);
+            convertNode->setDescs(inpEdge->getDesc(), outEdge->getDesc());
+            graph.InsertNode(inpEdge->getParent(), outEdge->getChild(), convertNode, iIndex, oIndex);
+
+            if (inpEdge->getDesc().getBlockingDesc() != outEdge->getDesc().getBlockingDesc()) {
+                //input and output layouts are different so the reorder is still needed
+                auto edge = convertNode->getChildEdgeAt(0);
+                auto convertOutputDesc = convertNode->getSelectedPrimitiveDescriptor()->getConfig().outConfs[0].desc;
+
+                std::string reorderName = edge->getParent()->getName() + "_" +
+                        MKLDNNExtensionUtils::getReorderArgs(convertOutputDesc, outEdge->getDesc()) +
+                        "_" + edge->getChild()->getName();
+                auto newReorder = graph.InsertReorder(edge, reorderName, convertOutputDesc, outEdge->getDesc());
+                if (newReorder) {
```

Comment:
In general, it is always a good idea to check a pointer that is returned from somewhere outside. Anyway this is not relevant anymore because the convert inserting procedure was moved to the mkldnn graph.